Move pkg/jsonlog to be a subpackage of the single consumer

dnephin opened the pull request:
Remove some unused code
Un-export some code that shouldn't be exported
Cleanup test cases
Move a constant to a more appropriate module
Move the pkg/jsonlog to a subpackage of jsonfilelog

- Part of #32989

dnephin committed:
Move RFC3339NanoFixed to a more appropriate package.

Signed-off-by: Daniel Nephin <dnephin@docker.com>

dnephin committed:
Unexport FastTimeMarshalJSON

Signed-off-by: Daniel Nephin <dnephin@docker.com>

dnephin committed:
Remove unused Format

Signed-off-by: Daniel Nephin <dnephin@docker.com>

dnephin committed:
Remove unused JSONLog marshaling

Signed-off-by: Daniel Nephin <dnephin@docker.com>

cpuguy83 reviewed (approved):
LGTM

cpuguy83 commented:
It would be nice to reconcile this with ffjson.

dnephin commented:
> It would be nice to reconcile this with ffjson.

I agree, but I think we need to get it setup with a `go generate`, and a check in `hack/validate` to make sure generation is done correctly, and a benchmark to compare against `encoding/json`.

unclejack commented:
@dnephin: Can you provide benchmark numbers for memory allocations done before and after these changes, please?

vdemeester commented on pkg/jsonlog/jsonlogbytes.go:
```diff
@@ -3,16 +3,15 @@ package jsonlog
-	Created string `json:"time"`
+	Log     []byte    `json:"log,omitempty"`
+	Stream  string    `json:"stream,omitempty"`
+	Created time.Time `json:"time"`
```
We are using that one on the client side right ? This might break a bunch of user of this `pkg` right ?

dnephin replied:
This struct is not used by the cli. The cli is only using a constant, and is made forward compatible here: https://github.com/docker/cli/pull/551

It might break users of this `pkg`, but that is unfortunately necessary. These packages were not maintained properly and developed really bad interfaces which need to be fixed.  We're going to be breaking everyone with #32989 anyway, and I think this can be seen as cleanup work necessary for #32989.

The godoc (https://godoc.org/github.com/moby/moby/pkg/jsonlog) doesn't show any consumers of this package. I'm not sure how reliable that is, but it seems to work for other packages.

dnephin commented:
Looking at this again, I don't think `JSONLog.MarshallJSON()` is actually used anywhere. It seems that object is only ever used to unmarshal JSON.

dnephin committed:
Fix benchmarks and remove more unnecessary code.

Signed-off-by: Daniel Nephin <dnephin@docker.com>

dnephin committed:
Move jsonlog to a subpackage of jsonfilelog

Signed-off-by: Daniel Nephin <dnephin@docker.com>

dnephin commented:
>  Can you provide benchmark numbers for memory allocations done before and after these changes

I found some benchmarks in `daemon/logger/jsonfilelog` and fixed them up to work, then backported them to master. Here's the output.

```
# Branch
go test -v -bench=. -run TestNothing -benchmem ./daemon/logger/jsonfilelog/
BenchmarkJSONFileLoggerLog-8             1000000              1007 ns/op    172.65 MB/s         121 B/op          1 allocs/op
BenchmarkJSONFileLoggerReadLogs-8        1000000              1243 ns/op    139.98 MB/s         356 B/op          6 allocs/op
PASS
ok      github.com/docker/docker/daemon/logger/jsonfilelog      2.284s

# Master
BenchmarkJSONFileLoggerLog-8             2000000               990 ns/op    175.61 MB/s         122 B/op          1 allocs/op
BenchmarkJSONFileLoggerReadLogs-8        1000000              1137 ns/op    153.01 MB/s         260 B/op          6 allocs/op
PASS
ok      github.com/docker/docker/daemon/logger/jsonfilelog      4.102s
```

This is the first go benchmark I've written, so let me know if you see anything that looks off. There were some logical errors in what exited before (wrong number of bytes because 1) it was using a different struct to calculate bytes, and 2) it was doing extra iterations not accounted for by `b.N`)

dnephin commented:
I removed some more unused code `JSONLog.MarshalJSON()` was completely unused, which allowed me to remove the duplcate ffjson function.  I moved the whole package to under `jsonfilelog` since this is really just an implementation detail of that logger.

vdemeester reviewed (approved):
LGTM 🐸

unclejack commented:
@dnephin: Thank you. I'm worried about regressions related to memory allocations in this code. We've had some unpleasant issues related to this in the past. It's good to see allocations remain pretty much the same.
